### ISSUE TITLE
[help wanted] experiment with autograd and coxph likelihood

### ISSUE THREAD

**Issue by CamDavidsonPilon** [CamDavidsonPilon/lifelines#655]
The current lifelines code for the Cox proportional hazard model's log-likelihood, gradient and hessian is all hand coded numpy + Python code. Specifically, the function `_get_efron_values_single` returns the log-likelihood, the gradient and the hessian. The latter two values are used in the Newton-Rhapson (also hand-crafted) to find the minimum. All this together has some problems:

1) There is still a Python for loop in the  `_get_efron_values_single` which bottlenecks performance
2) `_get_efron_values_single` is heavily optimized and hard to change
3) The hand-crafted Newton-Rhapson implementation may be sub optimal compared to scipy's `minimize`. 

**Experiment**: Can we use `autograd`  to create a log-likelihood and gradient calculation and `minimize` to find the optimal value that is **more** performant (clock time) than the current implementation? This is how we do it in the AFT models: [we provide `minimize` the autograd-generated gradients and value from the negative log-likehood](https://github.com/CamDavidsonPilon/lifelines/blob/d9d3f9f9acb832d03166e39556770c3374c868a4/lifelines/fitters/__init__.py#L1021-L1035). And the AFT models are much faster (mostly because there likelihoods are more Numpy-friendly I suspect).

Here's what I propose: 
1. Translate the Efron **log**-likelihood into pure Numpy (or _almost_ pure Numpy). The formula is [here](https://en.wikipedia.org/wiki/Proportional_hazards_model#Tied_times). You're free to use einsum, etc. to do this. 

2. Use `minimize` and `value_and_grad` to find the minimum value, and `hessian` to find the hessian matrix. Confirm that lifelines gives the same value. 

3. Time this implementation vs the current implementation. (I can help with this task).

**Comment by pr38** [comment]
Hi,

I came across this while trying to find a pure numpy solution for both the Hessian and Jacobian for the negative log-likehood loss. I don't know if it is something you are still interested.

Below is my code, but I should warn you that my code convergences at different weights than the lifelines package, but matches the weights of scikit-survival. I have yet to compare it against the packages in the R ecosystem. I had to use pytensor over autograd and pytorch, as it seems to be the only package that the do autograd over bincounts.


```
import numpy as np

from pytensor import tensor as pt
import pytensor

from lifelines.datasets import load_kidney_transplant
from lifelines import CoxPHFitter

from sksurv.linear_model import CoxPHSurvivalAnalysis
from sksurv.util import Surv

def get_neg_log_likelihood_loss_jacobian_hessian_function() -> pytensor.compile.function.types.Function:
    def reverse_cumsum(a):
        return pt.flip(pt.cumsum(pt.flip(a)))
    
    
    weights  = pt.vector('weights',dtype='float64')
    data = pt.matrix('data',dtype='float64')
    event = pt.vector('event',dtype='float64')
    
    n_samples = pt.scalar('n_samples',dtype='int64')
    n_unique_times = pt.scalar('n_unique_times',dtype='int64') 
    time_return_inverse = pt.vector('time_return_inverse',dtype='int64')
    
    p = pt.dot(data, weights)
    nums = pt.bincount(time_return_inverse,weights= p * event)
    risk_set = reverse_cumsum( pt.bincount(time_return_inverse,weights=pt.exp(p),minlength=n_unique_times ) )
    number_of_event_per_time = pt.bincount(time_return_inverse,weights=event.astype(np.int64) ,minlength=n_unique_times )
    neg_log_likelihood = pt.sum(-(nums - number_of_event_per_time * pt.log(risk_set)) / n_samples)
    
    jacobian = pytensor.gradient.jacobian(neg_log_likelihood,weights)
    hessian = pytensor.gradient.hessian(neg_log_likelihood,weights)
    neg_log_likelihood_loss_jacobian_hessian = pytensor.function(inputs=[weights,data,event,n_samples,n_unique_times,time_return_inverse],outputs= [neg_log_likelihood,jacobian,hessian])

    return neg_log_likelihood_loss_jacobian_hessian

data = load_kidney_transplant()

X = data[['age','black_male','white_male','black_female']].to_numpy()

def normalize(X):
    return np.subtract(X , X.mean(axis=0))/X.std(axis=0)

X = normalize(X)
time = data['time'].to_numpy()
event = data['death'].to_numpy()
unique_times, time_return_inverse =  np.unique(time,return_inverse=True)

weights = np.ones(X.shape[1])
step_size = 0.001

last_loss = np.array(np.inf)

for i in range(100):
    loss, jacobian, hessian = neg_log_likelihood_loss_jacobian_hessian(weights,X,event,X.shape[0],len(unique_times),time_return_inverse)
    if loss < last_loss:
        last_loss = loss
        weights = weights - (0.5 * np.dot(np.linalg.inv(hessian),jacobian))
    else:
        break

print(weights)

ll_chp = CoxPHFitter()
ll_chp.fit(data,'time','death')
print(ll_chp.summary)


sks = CoxPHSurvivalAnalysis()

sks.fit(X,Surv.from_arrays(event, time))

print(sks.coef_)
   


```

**Comment by CamDavidsonPilon** [comment]
Hi @pr38 
Your weights are probably different than lifelines (and I suspect Rs, too) but the same as sckit-survival because of how the different libraries treat ties. There are two methods, efron and breslow, of which the former is implemented in lifelines and R. Efron is more accurate, but computationally more difficult (hence it resisting my attempts to autograd it).

**Comment by pr38** [comment]
I believe you are referring to a heuristic I have seen a lot in the literature. Most people sort their data by survival time, and treat each row as a unique time, it makes computation easier. I believe the 'breslow' method is meant to deal with cases where many rows share the same survival times. My code above(and scikit-survival) calculates the 'full' loss (with the help of bincount). 

If you are interested, I can just sort and cumsum instead of using bincounts.

**Comment by CamDavidsonPilon** [comment]
> treat each row as a unique time, it makes computation easier

which is a tradeoff of accuracy vs computation. Maybe a solution is to jitter all times so they are always unique, run the fast computation many times to get many parameters, and take the average of those parameters. I've never tried this, but it's an interesting thought.
